naga/src/back/spv/mesh_shader.rs

```diff
@@ -0,0 +1,860 @@
+            let mut bi_index = 0;
+            let mut decorations = Vec::new();
+            for member in &mesh_return_info.vertex_info.struct_members {
+                if let crate::Binding::BuiltIn(_) = member.binding {
+                    ins.add_operand(member.ty_id);
+                    let binding = self.map_binding(
+                        ir_module,
+                        iface.stage,
+                        spirv::StorageClass::Output,
+                        // Unused except in fragment shaders with other conditions, so we can pass null
+                        Handle::new(NonMaxU32::new(0).unwrap()),
+                        &member.binding,
+                    )?;
+                    match binding {
+                        BindingDecorations::BuiltIn(bi, others) => {
+                            decorations.push(Instruction::member_decorate(
+                                builtin_block_ty_id,
+                                bi_index,
+                                spirv::Decoration::BuiltIn,
+                                &[bi as Word],
+                            ));
+                            for other in others {
+                                decorations.push(Instruction::member_decorate(
+                                    builtin_block_ty_id,
+                                    bi_index,
+                                    other,
+                                    &[],
+                                ));
+                            }
+                        }
+                        _ => unreachable!(),
+                    }
+                    bi_index += 1;
+                }
+            }
+            ins.to_words(&mut self.logical_layout.declarations);
+            decorations.push(Instruction::decorate(
+                builtin_block_ty_id,
+                spirv::Decoration::Block,
+                &[],
+            ));
+            for dec in decorations {
+                dec.to_words(&mut self.logical_layout.annotations);
+            }
+            let v =
+                self.write_mesh_return_global_variable(builtin_block_ty_id, vert_array_size_id)?;
+            iface.varying_ids.push(v);
+            if self.flags.contains(WriterFlags::DEBUG) {
+                self.debugs
+                    .push(Instruction::name(v, "naga_vertex_builtin_outputs"));
+            }
+            mesh_return_info.vertex_info.builtin_block = Some(v);
+        }
+        // Write primitive builtin block
+        if mesh_return_info
+            .primitive_info
+            .struct_members
+            .iter()
+            .any(|a| {
+                !matches!(
+                    a.binding,
+                    crate::Binding::BuiltIn(
+                        crate::BuiltIn::PointIndex
+                            | crate::BuiltIn::LineIndices
+                            | crate::BuiltIn::TriangleIndices
+                    ) | crate::Binding::Location { .. }
```

Comment:
Maybe these should go in an `impl crate::Binding` (in this module) block with a method name descriptive of the category they're in.